Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -122,16 +123,42 @@ internal sealed class RestChannel : ChannelBase
                 cancellationToken => SendAsync(restMethod, host, options, request, HttpCompletionOption.ResponseHeadersRead, cancellationToken));
 
             Task<Metadata> responseHeadersTask = ReadHeadersAsync(httpResponseTask);
-            Func<Status> statusFunc = () => GetStatus(ReadResponseAsync(httpResponseTask));
-            Func<Metadata> trailersFunc = () => GetTrailers(ReadResponseAsync(httpResponseTask));
-
-            PartialDecodingStreamReader<TResponse> responseStream = restMethod.ResponseStreamAsync<TResponse>(httpResponseTask, cancellationContext);
+            // The response will either contain the response (which is streaming, so we don't
+            // want to wait for it to finish) or error details (in which case we *do* want it
+            // to finish). The simplest way of providing status/trailer functionality is
+            // to create a separate ReadHttpResponseMessage with either empty content or
+            // the full original content. That's only used for status/trailers, and both
+            // GetStatus and GetTrailers only succeed when the response task has completed.
+            var statusResponseSource = new TaskCompletionSource<ReadHttpResponseMessage>();
+            var readerTask = CreateReaderTask();
+            Func<Status> statusFunc = () => GetStatus(statusResponseSource.Task);
+            Func<Metadata> trailersFunc = () => GetTrailers(statusResponseSource.Task);
+
+            PartialDecodingStreamReader<TResponse> responseStream = new PartialDecodingStreamReader<TResponse>(readerTask, restMethod.ParseJson<TResponse>, cancellationContext);
             Action disposalAction = () =>
             {
                 responseStream.Dispose();
                 cancellationContext.Cancel();
             };
             return new AsyncServerStreamingCall<TResponse>(responseStream, responseHeadersTask, statusFunc, trailersFunc, disposalAction);
+
+            async Task<TextReader> CreateReaderTask()
+            {
+                var httpResponse = await httpResponseTask.ConfigureAwait(false);
+                if (httpResponse.IsSuccessStatusCode)
+                {
+                    // The status for a successful response is empty, in terms of an error response.
+                    statusResponseSource.SetResult(new ReadHttpResponseMessage(httpResponse, ""));
```

Comment:
Would it be worth it to create a `StreamingHttpResponseMessage : ReadHttpResponseMessage` that exposes a Task<TextReader> StreamingContent as well as the existing Content inherited from ReadHttpResponseMessage. And you can encapsalute the logic of checking for status and setting StreamingContent or Content, etc.?